File: scrapers/kickass.js
const express = require("express");
const router = express.Router();
const axios = require("axios");
const cheerio = require("cheerio");
const BASE_URL = require("./constants").KICKASS_BASE_URL;
const isMagnet = require("./utils/misc_utils.js").isMagnet

router.get("/kickass_mg", async (req, res) => {
    var url = req.query.url;
    var response = await axios.get(url).catch(err=>{console.log(err)});
    var $ = cheerio.load(response.data);
    magnet = $("a.kaGiantButton").attr("href");
    if (isMagnet(magnet)) {
        res.status(200).json({ magnet: magnet }).end();
    } else {
        res.status(204).end();
    }
});

router.get("/kickass", async function (req, res) {
    //  Get the String to be Searched from URL
    const search = req.query.search;
    const html_response = true;
    const response = await axios.get(BASE_URL + search).catch((err) => {
        console.log(err);
        html_response = false;
    });
    const $ = cheerio.load(response.data);
    const selector = $("tr.odd , tr.even");
    const jsonResult = [];
    const content_avail = $("span[itemprop=name]").length;
    if (html_response) {
        if (content_avail == 0) {
            selector.each((i, element) => {
                name = $(element)
                    .find(".cellMainLink")
                    .text()
                    .replace("\n", "");
                uploader_name = $(element)
                    .find("td:nth-child(3)")
                    .text()
                    .replace("\n", "");
                file_size = $(element)
                    .find("td:nth-child(2)")
                    .text()
                    .replace("\n", "");
                upload_date = $(element)
                    .find("td:nth-child(4)")
                    .text()
                    .replace("\n", "");
                //  Seeders
                seeders = $(element)
                    .find("td:nth-child(5)")
                    .text()
                    .replace("\n", "");
                //  Leechers
                leechers = $(element)
                    .find("td:nth-child(6)")
                    .text()
                    .replace("\n", "");
                url =
                    "https://kickasstorrents.to" +
                    $(element).find(".cellMainLink").attr("href");
                jsonResult.push({
                    name: name,
                    torrent_url: url,
                    seeders: seeders,
                    leechers: leechers,
                    upload_date: upload_date,
                    size: file_size,
                    uploader: uploader_name,
                    magnet: "",
                    website: "Kickass",
                });
            });
            res.status(200).json({ data: jsonResult }).end();
        } else {
            res.status(204).end();
        }
    } else {
        res.status(204).end();
    }
});


module.exports = router;

File: routes/Kickass.go
package routes

import (
	"fmt"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gocolly/colly/v2"
)

func Kickass(ginCon *gin.Context) {
	search := strings.ReplaceAll(strings.TrimSpace(ginCon.Query("search")), " ", "%20")
	url := fmt.Sprint("https://kickasstorrents.to/usearch/", search)
	c := colly.NewCollector()
	infos := make([]TorrentInfo, 0)
	c.OnHTML("body", func(e *colly.HTMLElement) {
		ti := TorrentInfo{}
		if e.DOM.Find("span[itemprop=name]").Length() == 0 {
			e.ForEach("tr.odd , tr.even", func(i int, e *colly.HTMLElement) {
				ti.Name = e.ChildText(".cellMainLink")
				ti.Seeders = e.ChildText("td:nth-child(5)")
				ti.Leechers = e.ChildText("td:nth-child(6)")
				ti.Date = e.ChildText("td:nth-child(4)")
				ti.Size = e.ChildText("td:nth-child(2)")
				ti.Magnet = ""
				ti.Url = "https://kickasstorrents.to" + e.ChildAttr(".cellMainLink", "href")
				ti.Website = "Kickass"
				ti.Uploader = e.ChildText("td:nth-child(3)")
				ti.TorrentFileUrl = e.ChildAttr("td:nth-child(3) a:nth-child(2)", "href")
				infos = append(infos, ti)
			})
		}
	})
	c.OnError(func(r *colly.Response, err error) {
		fmt.Println("Request URL:", r.Request.URL, "failed with response:", r, "\nError:", err)
	})
	c.OnScraped(func(r *colly.Response) {
		if len(infos) > 0 {
			ginCon.JSON(200, TorrentRepo{infos})
		} else {
			ginCon.AbortWithStatus(204)
		}
	})
	c.Visit(url)
}

func Kickass_getMagnet(ginCon *gin.Context) {
	url := ginCon.Query("url")
	c := colly.NewCollector()
	var magnet, torrentfile string
	c.OnHTML("body", func(e *colly.HTMLElement) {
		magnet = e.ChildAttr("a.kaGiantButton", "href")
		torrentfile = getKickass_fileurl(magnet)
	})

	c.OnError(func(r *colly.Response, err error) {
		fmt.Println("Request URL:", r.Request.URL, "failed with response:", r, "\nError:", err)
	})

	c.OnScraped(func(r *colly.Response) {
		if strings.HasPrefix(magnet, "magnet") {
			ginCon.JSON(200, gin.H{"magnet": magnet, "torrentFile": torrentfile})
		} else {
			ginCon.AbortWithStatus(204)
		}
	})
	c.Visit(url)

}
